### ISSUE TITLE
Refactoring

### ISSUE THREAD

**Issue by tyiannak** [tyiannak/deep_audio_features#9]

**Comment by tyiannak** [comment]
Description: 
(copying @lobracost 's comments)

training.py:
After two bad (increasing loss), or neutral (non changing loss) epochs the training is finished. This choice might get you stucked at local minimas. A greater tolerance, for example a tolerance of 10 epoch,. would be better.

I suggest to use a learning rate scheduler (e.g. torch.optim.lr_scheduler.ReduceLROnPlateau) in order to optimize your training process.

test.py:
It would be useful if we had the ability to test a set of input data and not just a single file, so we could get some statistics.

I recommend to struct the code in a package form with init files. In this way there is no need to mess with the python path variable.

**Comment by lobracost** [comment]
The pkl directory, where the checkpoints are stored, is missing leading to post-training errors. We have to create this directory and commit the changes.

Also in the context of this refactoring, the unused imports and the commented code lines can be removed.

**Comment by tyiannak** [comment]
> The pkl directory, where the checkpoints are stored, is missing leading to post-training errors. We have to create this directory and commit the changes.
> 
> Also in the context of this refactoring, the unused imports and the commented code lines can be removed.

I've already done some unused imports removal and the pkl folder creation in task #8 (just pushed).

**Comment by DimitrisKatsiros** [comment]
@tyiannak @lobracost  

We can create a python package as suggested above and do all the checks (folders etc) in the `__init__()` or in the `setup.py`. 

Also, we can rename some files, for example: `basic_training.py` and `basic_test.py` to `training.py` and `test.py` respectively . 

We can add some _documentation / wiki_ pages, if the project is meant to be formal.

Finally I would suggest changing names of some variables, in order the code to be as formal as possible.

**Comment by tyiannak** [comment]
@lobracost 
I've started refactoring the code in this branch https://github.com/tyiannak/deep_audio_features/tree/Refactoring-%239, mainly focusing on 
 - reorganising submodules to folders so that imports can be made easier (so that the setup_paths.sh is not needed)
 - adding a setup file so that we can add the lib to pip
Can you please check that basic functionality is still ok both in command line scripts and through python calls?

You can also check with the current version of the pip binary (pip3 install deep-audio-features -U) which is based on this branch

**Comment by lobracost** [comment]
@tyiannak of course. I will let you know of the resulted behavior.

**Comment by tyiannak** [comment]
I've checked most of the functionality and it works (both with relative imports from the source code and using the pip install way), e.g.

from deep_audio_features.bin import basic_training as bt
bt.train_model(["/Users/tyiannak/Research/libraries/deep_audio_features/4class_small/music_small", "/Users/tyiannak/Research/libraries/deep_audio_features/4class_small/speech_small/"], "test")
...
from deep_audio_features.bin import basic_test as b
d, p = b.test_model("/Users/tyiannak/pkl/test.pt", "/Users/tyiannak/Research/libraries/deep_audio_features/r.wav", layers_dropped=0, test_segmentation=False)

but have not checked the combine functionalities at all (they certainly require changes)

**Comment by tyiannak** [comment]
@lobracost I've added everything (I think) in the main Readme
You can test with both command line (using the source code)
and in Python (using the latest pip version - aka 1.7)  - please test using both.
Also please add anything that you feel is missing from Readme wrt compare and then delete the compare/Readme file 

When u push these changes if any let me know to merge

**Comment by lobracost** [comment]
@tyiannak I want your opinion.

The combine.py script works as follows:

In order to extract nn features, it takes the output of the first convolutional layer (this choice is based on experimentation). These features are of shape (32, 64, 100). At the time the script functions as follows:

1. Squeezes the features to a 204800D vector.
2. Applies dimensionality reduction in order to produce a 100D vector

In my opinion, this is not efficient and has both performance issues and errors associated with memory handling (an array of size num_of_instances x 204800 has to be stored).

Do you have any better ideas in mind?

I propose simple averaging on axis 1 and 0 subsequently in order to get a 100d vector.

**Comment by tyiannak** [comment]
Two solutions a. Use a random projection to map to 100 features or b. Use features from the linear layers

**Comment by tyiannak** [comment]
I think b makes more sense

**Comment by tyiannak** [comment]
Or maybe some of the last convolutional mappings that will be lower in dimension

**Comment by lobracost** [comment]
I will experiment with both a and b and commit the changes for the best performing solution

**Comment by lobracost** [comment]
@tyiannak I experimented with both choices and concluded that the most effective way to extract features from this specific architecture is by taking the output of the fourth convolutional layer and applying dimensionality reduction. I committed my changes along with a minor update on the readme file.

All functionalities of the library work as expected for me, in both command line and python calls and thus I approve merging.
